include/sycl_blas_helper.h

```diff
@@ -32,6 +32,89 @@
 namespace blas {
 namespace helper {
 
+/**
+ * Allocation type for tests and benchmarks
+ */
+enum class AllocType : int { usm = 0, buffer = 1 };
+
+template <typename value_t, AllocType mem_alloc>
+struct AllocHelper;
+
+template <typename value_t>
+struct AllocHelper<value_t, AllocType::usm> {
+  using type = value_t *;
+};
+
+template <typename value_t>
+struct AllocHelper<value_t, AllocType::buffer> {
+  using type = blas::BufferIterator<value_t>;
+};
+
+#ifdef SB_ENABLE_USM
+template <AllocType alloc, typename value_t>
+typename std::enable_if<alloc == AllocType::usm,
+                        typename AllocHelper<value_t, alloc>::type>::type
+allocate(int size, cl::sycl::queue q) {
+  return cl::sycl::malloc_device<value_t>(size, q);
+}
+#endif
+
+template <AllocType alloc, typename value_t>
+typename std::enable_if<alloc == AllocType::buffer,
+                        typename AllocHelper<value_t, alloc>::type>::type
+allocate(int size, cl::sycl::queue q) {
+  return make_sycl_iterator_buffer<value_t>(size);
+}
+
+#ifdef SB_ENABLE_USM
+template <AllocType alloc, typename container_t>
+typename std::enable_if<alloc == AllocType::usm>::type deallocate(
+    container_t mem, cl::sycl::queue q) {
+  if (mem != NULL) {
+    cl::sycl::free(reinterpret_cast<void *>(mem), q);
+  }
+}
+#endif
+
+template <AllocType alloc, typename container_t>
+typename std::enable_if<alloc == AllocType::buffer>::type deallocate(
+    container_t mem, cl::sycl::queue q) {}
+
+// Need to add this guard since the enqueue_deallocate
+// function requires a host_task which throws a runtime
+// exception when used with the enable_profiling{} queue
+// property. We need to create all intermediate memory
+// before launching the kernel to avoid running into this
+// issue.
+// Enabling this code only for DEFAULT_CPU backend to get the
+// CI to pass.
+#ifdef DEFAULT_CPU
+template <AllocType alloc, typename container_t>
+typename std::enable_if<alloc == AllocType::usm, cl::sycl::event>::type
+enqueue_deallocate(std::vector<cl::sycl::event> dependencies, container_t mem,
+                   cl::sycl::queue q) {
+  cl::sycl::event event;
+#ifdef SB_ENABLE_USM
+  event = q.submit([&](cl::sycl::handler &cgh) {
+    cgh.depends_on(dependencies);
+    cgh.host_task([=]() { cl::sycl::free(mem, q); });
+  });
+#endif
+  return event;
+}
+
+template <AllocType alloc, typename container_t>
+typename std::enable_if<alloc == AllocType::buffer, cl::sycl::event>::type
+enqueue_deallocate(std::vector<cl::sycl::event>, container_t mem,
+                   cl::sycl::queue q) {
+  cl::sycl::event event;
+#ifdef SB_ENABLE_USM
+  event = q.submit([&](cl::sycl::handler &cgh) { cgh.host_task([=]() {}); });
```

Comment:
What is the reason of spawning a `host_task` that does nothing? Can we remove it?